tidy3d/components/data/near2far_data.py

```diff
@@ -0,0 +1,732 @@
+        """
+
+        # Assumes that frequencies and angles are the same for all radiation vectors
+        frequencies = np.array(self.monitor.freqs)
+        ux = np.array(self.monitor.ux)
+        uy = np.array(self.monitor.uy)
+
+        eta = np.array([self.eta(frequency, medium) for frequency in frequencies])
+
+        # assemble E felds
+        eta = eta[None, None, ...]
+        Et_array = -(self.Lphi.values + eta * self.Ntheta.values)
+        Ep_array = self.Ltheta.values - eta * self.Nphi.values
+        Er_array = np.zeros_like(Ep_array)
+
+        # assemble H fields
+        Ht_array = -Ep_array / eta
+        Hp_array = Et_array / eta
+        Hr_array = np.zeros_like(Hp_array)
+
+        dims = ("ux", "uy", "f")
+        coords = {"ux": ux, "uy": uy, "f": frequencies}
+
+        Er = xr.DataArray(data=Er_array, coords=coords, dims=dims)
+        Et = xr.DataArray(data=Et_array, coords=coords, dims=dims)
+        Ep = xr.DataArray(data=Ep_array, coords=coords, dims=dims)
+
+        Hr = xr.DataArray(data=Hr_array, coords=coords, dims=dims)
+        Ht = xr.DataArray(data=Ht_array, coords=coords, dims=dims)
+        Hp = xr.DataArray(data=Hp_array, coords=coords, dims=dims)
+
+        field_data = xr.Dataset(
+            {"E_r": Er, "E_theta": Et, "E_phi": Ep, "H_r": Hr, "H_theta": Ht, "H_phi": Hp}
+        )
+
+        return field_data
+
+    def power(self, medium: Medium = Medium(permittivity=1)) -> xr.Dataset:
+        """Get power on the observation grid defined in k-space.
+
+        Parameters
+        ----------
+        medium : :class:`.Medium`
+            Background medium in which to radiate near fields to far fields.
+            Default: free space.
+
+        Returns
+        -------
+        power_data: xarray.Dataset
+            xarray dataset containing power.
+        """
+
+        # Assumes that frequencies and angles are the same for all radiation vectors
+        frequencies = np.array(self.monitor.freqs)
+        ux = np.array(self.monitor.ux)
+        uy = np.array(self.monitor.uy)
+
+        eta = np.array([self.eta(frequency, medium) for frequency in frequencies])
+
+        # assemble E felds
+        eta = eta[None, None, ...]
+        Et_array = -(self.Lphi.values + eta * self.Ntheta.values)
+        Ep_array = self.Ltheta.values - eta * self.Nphi.values
+
+        # assemble H fields
+        Ht_array = -Ep_array / eta
+        Hp_array = Et_array / eta
+
+        power_theta = 0.5 * np.real(Et_array * np.conj(Hp_array))
+        power_phi = 0.5 * np.real(-Ep_array * np.conj(Ht_array))
+        values = power_theta + power_phi
+
+        dims = ("ux", "uy", "f")
+        coords = {"ux": ux, "uy": uy, "f": frequencies}
+
+        power_data = xr.DataArray(data=values, coords=coords, dims=dims)
+
+        return power_data
+
+
+Near2FarDataTypes = (Near2FarAngleData, Near2FarCartesianData, Near2FarKSpaceData)
+Near2FarDataType = Union[Near2FarDataTypes]
+DATA_TYPE_MAP = DATA_TYPE_MAP.update(
```

Comment:
I had it that way at the start but it gave errors in the backend test, e.g.:
```    sim_data = SimulationData(
File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for SimulationData
monitor_data -> n2f_angle
 No match for discriminator 'type' and value 'Near2FarAngleData' (allowed values: 'FieldData', 'FieldTimeData', 'PermittivityData', 'ModeSolverData', 'ModeData', 'FluxData', 'FluxTimeData') (type=value_error.discriminated_union.invalid_discriminator; discriminator_key=type; discriminator_value=Near2FarAngleData; allowed_values='FieldData', 'FieldTimeData', 'PermittivityData', 'ModeSolverData', 'ModeData', 'FluxData', 'FluxTimeData')```

I added `MonitorDataType = Union[MonitorDataType, Near2FarDataType]` below the line you marked, but maybe that's not the right way to do it?

Reply:
so these lines won't get run until the module is imported I believe.  So if someone imports `SimulationData` from `sim_data.py`, it will try importing `MonitorDataType` from `.monitor_data`. Maybe adding the `MonitorDataType = Union[MonitorDataType, Near2FarDataType]` to `near2far_data.py` never gets run because that module isn't imported.   I wonder if the best is to import everything into `data/__init__.py` and then construct the `MonitorDataType` and `DATA_TYPE_MAP` directly in there.  My worry is that if someone imports `MonitorDataType` from `monitor_data.py`, it might be different from importing it from `sim_data.py`. We should make sure it's consistent.

Reply:
Sorry, I'm not sure I understood the suggestion. Right now, in `near2far_data.py` we have

```python
class AbstractNear2FarData(MonitorData): # needs to import from monitor_data.py
...
class Near2FarAngleData(AbstractNear2FarData):
...
class Near2FarCartesianData(AbstractNear2FarData):
...
class Near2FarKSpaceData(AbstractNear2FarData):
...
Near2FarDataTypes = (Near2FarAngleData, Near2FarCartesianData, Near2FarKSpaceData)
Near2FarDataType = Union[Near2FarDataTypes]
```
In order to define the union `Near2FarDataType` in `monitor_data.py`, we'd need to somehow avoid the circular import.
If I interpret your suggestion correctly: we create a new file `near2far_utils.py` which has:
```python
from near2far_data import Near2FarDataTypes
Near2FarDataType = Union[Near2FarDataTypes]
```
then import this file into `monitor_data.py` and update `MonitorDataType` there? Isn't this still a circular import, just indirect?

Another option is to make a more general `utils.py` file in the `data` folder, containing
```python
from monitor_data import MonitorDataTypes
from near2far_data import Near2FarDataTypes

MonitorDataType = Union[MonitorDataTypes, Near2FarDataTypes]
```
and then in `sim_data.py` put `from utils import MonitorDataType`?

Or did you mean we should move the actual classes from `near2far_data.py` into `monitor_data.py`?